src/ethereum/frontier/trie.py

```diff
@@ -13,21 +13,209 @@
+
+
+@slotted_freezable
+@dataclass
+class BranchNode:
+    """Branch node in the Merkle Trie"""
+
+    subnodes: List[rlp.RLP]
+    value: rlp.RLP
+
+
+InternalNode = Union[LeafNode, ExtensionNode, BranchNode]
+
+
+def encode_internal_node(node: Optional[InternalNode]) -> rlp.RLP:
+    """
+    Encodes a Merkle Trie node into its RLP form. The RLP will then be
+    serialized into a `Bytes` and hashed unless it is less that 32 bytes
+    when serialized.
+
+    This function also accepts `None`, representing the absence of a node,
+    which is encoded to `b""`.
+
+    Parameters
+    ----------
+    node : Optional[InternalNode]
+        The node to encode.
+
+    Returns
+    -------
+    encoded : `rlp.RLP`
+        The node encoded as RLP.
+    """
+    unencoded: rlp.RLP
+    if node is None:
+        unencoded = b""
+    elif isinstance(node, LeafNode):
+        unencoded = (
+            nibble_list_to_compact(node.rest_of_key, True),
+            node.value,
+        )
+    elif isinstance(node, ExtensionNode):
+        unencoded = (
+            nibble_list_to_compact(node.key_segment, False),
+            node.subnode,
+        )
+    elif isinstance(node, BranchNode):
+        unencoded = node.subnodes + [node.value]
+    else:
+        raise Exception(f"Invalid internal node type {type(node)}!")
+
+    encoded = rlp.encode(unencoded)
+    if len(encoded) < 32:
+        return unencoded
+    else:
+        return crypto.keccak256(encoded)
+
+
+def encode_node(node: Node, storage_root: Optional[Bytes] = None) -> Bytes:
+    """
+    Encode a Node for storage in the Merkle Trie.
+
+    Currently mostly an unimplemented stub.
+    """
+    if isinstance(node, Account):
+        assert storage_root is not None
+        return rlp.encode_account(node, storage_root)
+    elif isinstance(node, (Transaction, Receipt)):
+        return rlp.encode(cast(rlp.RLP, node))
+    else:
+        return cast(Bytes, node)
+
+
+@dataclass
+class Trie:
+    """
+    The Merkle Trie.
+    """
+
+    secured: bool
+    _data: Dict[Bytes, Node]
+
+    def __init__(self, secured: bool) -> None:
+        self.secured = secured
+        self._data = {}
+
+
+def trie_set(trie: Trie, key: Bytes, value: Node) -> None:
+    """
+    Stores an item in a Merkle Trie.
+
+    This method deletes the key if the value is `b""`, since the Merkle
+    Trie does not distinguish between missing and `b""`.
+
+    Parameters
+    ----------
+    trie: `Trie`
+        Trie to store in.
+    key : Bytes
+        Key to lookup.
+    value :
+        Node to insert at `key`.
+    """
+    if value in (EMPTY_ACCOUNT, U256(0), b""):
+        if key in trie._data:
+            del trie._data[key]
+    else:
+        trie._data[key] = value
+
+
+def trie_get(trie: Trie, key: Bytes) -> Node:
+    """
+    Gets an item from the Merkle Trie.
+
+    This method returns `b""` if the key is missing, since the Merkle Trie
+    does not distinguish between missing and `b""`.
```

Comment:
Your proposed code is complete nonsense. Since annotations are simply hints that allow `pypy` to do typechecking and aren't used at runtime. `type(N) == TypeVar` and `N` isn't even callable.

I've taken the straightforward approach of giving `trie_get()` a `default` parameter with the same semantics as `dict.get()`.

Reply:
> Your proposed code is complete nonsense. Since annotations are simply hints that allow `pypy` [sic] to do typechecking and aren't used at runtime. `type(N) == TypeVar` and `N` isn't even callable.

This is a little harsh, especially since I noted _in my comment_ that you can't directly call a `TypeVar`. I was quite aware the sample wouldn't run as written, but I figured the spirit of what I suggested was clear enough: pass in the type of value stored in the trie, and default construct it when the key doesn't exist. Here's a more fleshed out version:

```python
N = TypeVar("N", Account, Bytes, Transaction, Receipt, Uint, U256)

def trie_get(class_: Type[N], trie: Trie[N], key: Bytes) -> N:
    try:
        return trie._data[key]
    except KeyError:
        return class_()
```

**NB**: I haven't run this through `mypy`, and would need to check that the `Type[N]` annotation behaves as I expect.

> I've taken the straightforward approach of giving `trie_get()` a `default` parameter with the same semantics as `dict.get()`.

The upside to adding a generic argument to `Trie` is that we get type safety between the different kinds of tries. You can't accidentally put a `Bytes` in the root trie (only `Account`), and vice versa. I understand that might be a bit more work, though, for a tiny payoff.

Whether we pass in a `Callable` or a default value into `trie_get` is somewhat immaterial to me, since constructing a dataclass should be a relatively lightweight operation.

Reply:
> This is a little harsh, especially since I noted in my comment that you can't directly call a TypeVar.

Sorry for being thoughtlessly rude.

> Whether we pass in a `Callable` or a default value into `trie_get` is somewhat immaterial to me, since constructing a dataclass should be a relatively lightweight operation.

I think it's simpler to store the default value in the `Trie` object. This is incompatible with Verkle Tries, when both accounts and storage will be placed in the same trie, but that is a bridge to cross when we come to it.